### PR TITLE
Add corpus + tests/benchmarking for lexing

### PR DESCRIPTION
**Pull request by colinwahl**
This PR adds a `corpus` directory, which will contain the set of pact/repl files we'd like to run tests/benchmarks over.

For now, the `corpus` directory is a clone of [pact-examples](https://github.com/kadena-io/pact-examples) with the non-pact/repl files ripped out.

This PR also adds two proc macros which generate tests & benchmarks for each file in the corpus directory. This is done via reading the files into a tree data structure and then walking the tree, generating the appropriate rust code to handle setting up tests/benchmarks. That work was based off of [test-each-file](https://github.com/binary-banter/test-each-file) but specialized to our needs. To do this, I brought in some low level macro code by dtolnay which makes writing macros a lot easier.

The good news is that our current lexer implementation does lex each file in the corpus correctly (including repl files!). By "correctly", I mean that the token stream doesn't contain any error nodes.

These benchmarks will allow us to keep track of our performance over each corpus file as we develop so that we can have a realistic view of how our changes affect performance on "real" code.

### CODE REVIEW

**Comment by colinwahl** [comment]
Here's what the generated test output looks like:

```
test lexer::tests::test_corpus::pact_examples::accounts::accounts_pact ... ok
test lexer::tests::test_corpus::pact_examples::atoz::_01_keysets::keysets_pact ... ok
test lexer::tests::test_corpus::pact_examples::atoz::_01_keysets::keysets_repl ... ok
test lexer::tests::test_corpus::pact_examples::atoz::_02_module::auth_pact ... ok
test lexer::tests::test_corpus::pact_examples::commercial_paper::auth_pact ... ok
test lexer::tests::test_corpus::pact_examples::accounts::accounts_repl ... ok
test lexer::tests::test_corpus::pact_examples::atoz::_02_module::auth_repl ... ok
test lexer::tests::test_corpus::pact_examples::commercial_paper::cp_pact ... ok
test lexer::tests::test_corpus::pact_examples::commercial_paper::cp_repl ... ok
test lexer::tests::test_corpus::pact_examples::crypto_critters::critters_repl ... ok
test lexer::tests::test_corpus::pact_examples::commercial_paper::orders_pact ... ok
test lexer::tests::test_corpus::pact_examples::hello_world::hello_world_pact ... ok
test lexer::tests::test_corpus::pact_examples::hello_world::hello_world_repl ... ok
test lexer::tests::test_corpus::pact_examples::pact3_0::_02_coin_cap_repl ... ok
test lexer::tests::test_corpus::pact_examples::pact3_0::_01_coin_gov_repl ... ok
test lexer::tests::test_corpus::pact_examples::pact3_0::_03_coin_interface_repl ... ok
test lexer::tests::test_corpus::pact_examples::verified_accounts::accounts_repl ... ok
test lexer::tests::test_corpus::pact_examples::verified_accounts::accounts_pact ... ok
```

File: corpus/pact-examples/pact3.0/03-coin-interface.repl
Comment (thomashoneyman) [inline corpus/pact-examples/pact3.0/03-coin-interface.repl:1]: I think we should remove the pact-3.0 examples from the corpus; they’re from an older version of the language and a lot of the content looks similar to the later contracts (like coin)

File: macros/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "pactfmt-macros"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+proc-macro = true
+
+[dependencies]
+quote = "1"
```
Comment (thomashoneyman) [inline macros/Cargo.toml:10]: What does "1" mean here? Is it exactly 1.0.0? Why not SemVer?

File: macros/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+}
+
+#[proc_macro]
+pub fn test_corpus(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let args = parse_macro_input!(input as TestCorpusArgs);
+
+    let mut tokens = TokenStream::new();
+
+    let corpus = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/../corpus"));
```
Comment (thomashoneyman) [inline macros/src/lib.rs:131]: What is CARGO_MANIFEST_DIR here?

File: macros/Cargo.toml (hunk 1 above)
Comment (colinwahl) [reply macros/Cargo.toml:10]: "1" means ">=1.0.0, <2.0.0" https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html

However, I was just being lazy. I really ought to have specified an exact version. I'll do that.

Comment (thomashoneyman) [reply macros/Cargo.toml:10]: I don’t mind — we can do it that way, just curious!

**Review by thomashoneyman — APPROVED** [review]

File: macros/Cargo.toml (hunk 1 above)
Comment (colinwahl) [reply macros/Cargo.toml:10]: Ok! In that case I'll leave it for now

**Pull request merged**